File: 215/215.js
// max heap solution



// min heap solution






// n + n/2 + n/4 + n/8 + ... + 1 = x
// 2n + n + n/2 + n/4 + ... + 2 = 2x
// 2n + 2 = x

//O(n)



// when i did this problem again i did it iteratively. it got TLE and it took me a long time to realize that i forgot to update the left and right boundaries
// in my iterative version.

var findKthLargest2 = function(nums, k) {
    let count = 0;
    function swap(a, b) {
        let temp = nums[a];
        nums[a] = nums[b];
        nums[b] = temp;
    }
    
    function partition(l, r) {
        count++;
        let pivot = nums[r];
        let i =l;
        for(let j=l; j<r;j++) {
            if(nums[j]>pivot){
                swap(i, j);
                i++;
            }
        }
        swap(i, r);
        return i;
    }
    let l = 0;
    let r = nums.length-1;
    let i = partition(l, r);
    while(true) {
        if(i === k-1) {
            return nums[i];
        } else if(k-1>i) {
            l = i+1;
            i = partition(l, r);
        } else {
            r = i-1;
            i = partition(l, r);
        }
    }
};




var findKthLargest = function(nums, k) {
    const swap = (a, b) => {
        let temp = nums[a];
        nums[a] = nums[b];
        nums[b] = temp;
    }
    const partition = (l, r) => {
        let pivot = nums[r];
        let i = l;
        for(let j=l; j<r;j++){
            if(nums[j]>pivot) {
                swap(i, j);
                i++;
            } 
        }
        swap(i, r);
        return i;
    }
    
    const search = (l, r) => {
        
        let i = partition(l, r);

        if(i === k-1) {
            
            return nums[i];
        } else if(k-1>i) {
            
            return search(i+1, r);
        } else {
            
            return search(l, i-1);
        }
    }
    let l = 0;
    let r = nums.length-1;
    let res = search(l,r);
    return res;
    
};

File: 215/215.py
# 215. Kth Largest Element in an Array

# heap
class Solution:
    def findKthLargest(self, nums: List[int], k: int) -> int:
        nums = list(map(lambda x:-x, nums))
        heapq.heapify(nums)
        ans = None
        for i in range(k):
            ans=heapq.heappop(nums)
        return -ans


# quick select
# this tle for 1 test case
class Solution:
    def findKthLargest(self, nums: List[int], k: int) -> int:
        def partition(start,end,arr):
            pivot = arr[end]
            left = start
            for right in range(start, end):
                if arr[right] <= pivot:
                    arr[left], arr[right] = arr[right], arr[left]
                    left+=1
            arr[left], arr[end] = arr[end], arr[left]
            return left
        s=0
        e=len(nums)-1
        while True:
            p = partition(s,e,nums)
            if p==len(nums)-k:
                return nums[p]
            if p<len(nums)-k:
                s=p+1
            else:
                e=p-1

# O(n) space. Not in place, but helps with repeating numbers
# average O(n) time complexity. n+n/2+n/4+...
# worst case O(n^2) time complexity

class Solution:
    def findKthLargest(self, nums: List[int], k: int) -> int:
        def partition(arr):
            pivot=arr[0]
            left,mid,right=[],[],[]
            for a in arr:
                if a<pivot:
                    left.append(a)
                elif a>pivot:
                    right.append(a)
                else:
                    mid.append(a)
            return (left, mid, right)
          
        def find(arr, j):
            l,m,r = partition(arr)
            if j<=len(r):
                return find(r, j)
            if j>len(r)+len(m):
                return find(l,j-len(r)-len(m))
            return m[0]
        
        return find(nums,k)
